teps/0040-capture-step-exit-code.md

```diff
@@ -0,0 +1,70 @@
+---
+status: proposed
+title: 'Capture Step Exit Code'
+creation-date: '2021-01-06'
+last-updated: '2021-01-06'
+authors:
+- '@pritidesai'
+- '@afrittoli'
+---
+
+# TEP-0040: Capture Step Exit Code
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Tekton tasks are defined as a collection of steps in which each step can specify a container image to run.
+Steps are executed in order in which they are specified. One single step failure results in a task failure
+in turn results in a pipeline failure. Once a step results in failure, rest of the steps are not executed.
+
+Many common pipelines have requirement where a step failure must not halt the entire execution. In order to
+continue executing a task even if a step fail, pipeline authors have flexibility of wrapping an image and
+exiting that task with success. This changes the failing task into success and does not block further
+execution. This is a workaround and only works with images which can be wrapped. This workaround does not apply
+to off-the-shelf container images.
+
+As a pipeline execution engine, we want to support off-the-shelf container image as a step and provide
+an option to capture non-zero exit code. The pipeline author can choose to continue execution after capturing
+the non-zero exit code and make it available to access it after the execution.
+
+Issue: [tektoncd/pipeline#2800](https://github.com/tektoncd/pipeline/issues/2800)
+
+
+## Motivation
+
+It should be possible to easily use off-the-shelves (OTS) images as steps in Tekton tasks. A task author has no
+control on the image but may desire to capture an error and continue executing the pipeline.
+
+### Goals
+
+Design a step failure strategy so that the task author can control the behaviour of the image and decide to
+continue execution or not.
+
+Prevent a task from failing when a step fails.
+
+Store the container termination state or error state and make it accessible after the pipeline finishes execution.
+
+This proposal must be applicable to any container image including custom images and off-the-shelf images.
+
+### Non-Goals
+
+This design is limited to a step within a task and does not apply to pipeline tasks.
+
+### Use Cases
+
+As a pipeline author, I would like to design a task with multiple steps. One of the steps is running an
+enterprise  image to run unit tests, and the next step needs to report test results even after a previous
+step results in failure due to tests failure.
+
```

Comment:
Can you go into more detail with the use cases?

Reply:
One example could be the nightly release pipeline. As part of the release pipeline we might want to run a security scan, unit tests and other checks. None of this checks should stop the release pipeline from executing, however we want to run the tests so that we keep a track of which failures we saw against which version.

This can be sometime achieved by wrapping the step that runs the tests with a script, but in other cases it would require to create a new container image that contains a shell. In cases where wrapping with a script is possible, it may still be undesirable.

This use case is strengthened by the case of users migrating to Tekton. They may have scripts / steps from their previous CI system where it was possible to ignore a failure, and they would have to modify them and/or wrap them to trap the exit code.

Reply:
Having to modify a `Task` is the main downside of this approach. 
The use case I'm facing is to migrate users to Tekton, they already have tasks written, and they need to be moved to Tekton, so it is not an issue in this case.

A possible workaround for existing Tasks, they could expose the capture flag through a task parameter, to allow the Task to use in both modes, with and without capture.

Reply:
"we'd have to modify the golang-test task" ... wow I somehow hadn't considered that. All of our legacy cases are with teams providing an "image" that when run might "fail", but in Tekton where we're promoting a Task Catalog teams might provide a Task that we should equivalently let a Pipeline decide if it should fail. I think maybe we should consider support ignoring Task failure in this TEP or at least immediately create another TEP to look at adding support.